File: sdk/python/examples/08_camera_record.py
"""录像示例：录制指定时长的视频。

用法:
    # 录制 10 秒 (默认)
    python 08_camera_record.py

    # 录制 30 秒
    python 08_camera_record.py 30

    # 指定摄像头源和时长
    python 08_camera_record.py 15 rtsp://192.168.66.190:8554/cam
"""

import sys
import time

from brainstem_sdk.camera import OrixCamera

# ── 参数解析 ────────────────────────────────────────────────

duration = 10.0  # 默认录制 10 秒
source = 0       # 默认 USB 摄像头

if len(sys.argv) > 1:
    try:
        duration = float(sys.argv[1])
    except ValueError:
        source = sys.argv[1]

if len(sys.argv) > 2:
    arg = sys.argv[2]
    source = int(arg) if arg.isdigit() else arg

# ── 录制 ────────────────────────────────────────────────────

output_file = "orix_recording.mp4"

with OrixCamera(source=source) as cam:
    w, h = cam.get_frame_size()
    print(f"摄像头: {w}x{h}")
    print(f"开始录制 {duration} 秒 → {output_file}")
    print("录制中... (按 Ctrl+C 提前停止)")

    cam.start_recording(output_file, duration=duration)

    start = time.monotonic()
    try:
        while cam.is_recording:
            elapsed = time.monotonic() - start
            remaining = max(0.0, duration - elapsed)
            print(
                f"\r  已录制 {elapsed:.1f}s / {duration:.1f}s "
                f"(剩余 {remaining:.1f}s)",
                end="", flush=True,
            )
            time.sleep(0.5)
    except KeyboardInterrupt:
        print("\n用户中断，停止录制...")
        cam.stop_recording()

    print(f"\n录制完成: {output_file}")

File: sdk/cpp/examples/08_camera_record.cpp
// 08_camera_record.cpp — 后台录像：定时录制 + 手动停止
#include "orix/orix_camera.h"

#include <chrono>
#include <csignal>
#include <iostream>
#include <thread>

static volatile bool g_interrupt = false;
static void sig_handler(int) { g_interrupt = true; }

int main(int argc, char* argv[]) {
    double duration = 10.0; // 默认录制 10 秒
    if (argc >= 2) duration = std::stod(argv[1]);

    std::signal(SIGINT, sig_handler);

    try {
        orix::OrixCamera cam(0, 640, 480, 30);
        cam.open();

        auto [w, h] = cam.get_frame_size();
        std::cout << "摄像头: " << w << "x" << h << "\n";

        // ── 定时录制 ──────────────────────────────────────────
        std::cout << "开始录制（" << duration << " 秒）: orix_record.mp4\n";
        cam.start_recording("orix_record.mp4", duration);

        auto t0 = std::chrono::steady_clock::now();
        while (cam.is_recording() && !g_interrupt) {
            auto elapsed = std::chrono::duration<double>(
                std::chrono::steady_clock::now() - t0).count();
            std::cout << "\r录制中... " << std::fixed << std::setprecision(1)
                      << elapsed << " / " << duration << " s  " << std::flush;
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
        }
        std::cout << "\n";

        if (g_interrupt) {
            std::cout << "收到中断信号，停止录制...\n";
            cam.stop_recording();
        } else {
            // 等录制线程自然结束
            while (cam.is_recording())
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        std::cout << "录制完成: orix_record.mp4\n";

        // ── 验证：读取录制文件的帧数 ──────────────────────────
        cv::VideoCapture verify("orix_record.mp4");
        if (verify.isOpened()) {
            int frames = static_cast<int>(verify.get(cv::CAP_PROP_FRAME_COUNT));
            double fps = verify.get(cv::CAP_PROP_FPS);
            std::cout << "验证: " << frames << " 帧, "
                      << fps << " fps, "
                      << (frames / fps) << " 秒\n";
        }

        cam.close();

    } catch (const std::exception& e) {
        std::cerr << "错误: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
